### PR TITLE
pre-commit hooks update; mypy, ruff fixes

### PR DESCRIPTION
**Pull request by martinhoyer**
New mypy version was throwing error. I've done my best to address them, but definitely needs @happz's eyes.

Refurb rules are now out of 'preview' and surprisingly, tmt passes almost everything with flying colors :rocket:.   
Only thing it didn't like regex flag aliases and readlines(), which was used only in one place.

Not updating to latest version of pyright. There are errors linked to UnserializeCallback, which I couldn't properly figure out. 
```
T = TypeVar('T')
UnserializeCallback = Callable[[Any], T]
```
FWIW - from https://github.com/microsoft/pyright/releases 1.1.372:

> Fixed bug that results in some circumstances in incorrect specialization of type[T] when T evaluates to Any.
> Fixed false positive error when using an overloaded __init_subclass__ method under certain circumstances.

### CODE REVIEW

File: tmt/utils/__init__.py [hunk 1]
```diff
@@ -2158,7 +2158,7 @@ class _MultiInvokableCommonMeta(_CommonMeta):
     """
 
     # N805: ruff does not recognize this as a metaclass, `cls` is correct
-    def __init__(cls, *args: Any, **kwargs: Any) -> None:  # noqa: N805
+    def __init__(cls, *args: Any, **kwargs: Any) -> None:
```
Comment (happz) [inline tmt/utils/__init__.py:2161]: Comment above probably no longer applies

File: tmt/utils/__init__.py [hunk 2]
```diff
@@ -2965,14 +2949,8 @@ def container_field(
...
         metadata = field.metadata.get('tmt', FieldMetadata())
-        if inspect.isclass(container):
-            return field.name, key_to_option(field.name), field, metadata
-
-        return (
-            field.name,
-            key_to_option(field.name),
-            container.__dict__[field.name],
-            field, metadata)
+        fname = container.__dict__[field.name] if not inspect.isclass(container) else None
```
Comment (happz) [inline tmt/utils/__init__.py:2952]: Is `fname` a fitting name? It looks more like a value of the field.

File: tmt/steps/prepare/__init__.py [hunk 3]
```diff
@@ -223,7 +224,8 @@ def as_key(self) -> frozenset['tmt.base.DependencySimple']:
                     continue
 
                 collected_requires[guest].dependencies += tmt.base.assert_simple_dependencies(
-                    phase.essential_requires(),
+                    phase.essential_requires(),  # type: ignore[attr-defined]
+                    # mypy thinks that phase is Phase type, while its actually PluginClass
```
Comment (happz) [inline tmt/steps/prepare/__init__.py:228]: Nitpicking: can you put the comment *before* the line with the waiver, and include the waiver? Not often, but it sometimes happen there are more issues to be waived, and then it's nice to have a comment for each of them:

```python
# ignore[attr-defined]: mypy thinks that phase is Phase type, while its actually PluginClass
phase.essential_requires(),  # type: ignore[attr-defined]
```

File: tmt/utils/__init__.py (hunk 2 above)
Comment (martinhoyer) [reply tmt/utils/__init__.py:2952]: I'm super bad at naming things. What would you suggest?

Comment (happz) [reply tmt/utils/__init__.py:2952]: `value`, or maybe not even assign it to a dedicated variable & return it right away:

```python
return (
    field.name,
    key_to_option(field.name),
    container.__dict__[field.name] if not inspect.isclass(container) else None,
    field,
    metadata)

File: tmt/utils/__init__.py (hunk 1 above)
Comment (martinhoyer) [reply tmt/utils/__init__.py:2160]: addressed in https://github.com/teemtee/tmt/pull/3142/commits/feeabd484f3a97e109a126f465d56cebddda694c

File: tmt/utils/__init__.py (hunk 2 above)
Comment (martinhoyer) [reply tmt/utils/__init__.py]: Good point, thanks, edited in https://github.com/teemtee/tmt/pull/3142/commits/d0acfc110faca3a70b0d8d8cc1987fd2aa1fd6dc

File: tmt/steps/prepare/__init__.py (hunk 3 above)
Comment (martinhoyer) [reply tmt/steps/prepare/__init__.py:227]: Addressed in https://github.com/teemtee/tmt/pull/3142/commits/feeabd484f3a97e109a126f465d56cebddda694c

**Review by happz — APPROVED** [review]

File: tmt/steps/prepare/__init__.py [hunk 4]
```diff
@@ -187,7 +187,8 @@ class DependencyCollection:
 
             @property
             def as_key(self) -> frozenset['tmt.base.DependencySimple']:
-                return frozenset(collection.dependencies)
+                return frozenset(collection.dependencies)  # type: ignore[has-type]
+                # mypy doesn't seem to understand collection as `frozenset`
```
Comment (thrix) [inline tmt/steps/prepare/__init__.py:191]: same nitpick as from @happz , comment ideally before :)

File: tmt/steps/prepare/distgit.py [hunk 5]
```diff
@@ -287,11 +287,11 @@ def go(
             self.discover.post_dist_git(created_content)
             # FIXME needs refactor of Prepare, tmt.base etc...
             # doing quick & dirty injection of prepareinstalls
-            for guest in self.step.plan.provision.guests():
+            for g in self.step.plan.provision.guests():
```
Comment (thrix) [inline tmt/steps/prepare/distgit.py:290]: @martinhoyer was this reported by ruff? I personally rather prefer `guest` here.

**Comment by seberm** [comment]
Please remember to rebase your changes on top of the `main` just before merging to make sure all the code passes the checks.

**Review by thrix — APPROVED** [review]
LGTM (with some small comments)

**Comment by happz** [comment]
> Please remember to rebase your changes on top of the `main` just before merging to make sure all the code passes the checks.

No worries, he who merges is always limited to fast-forward only, therefore rebase to `main` is a must.

File: tmt/steps/prepare/__init__.py (hunk 3 above)
Comment (happz) [reply tmt/steps/prepare/__init__.py:227]:
```suggestion
                    # ignore[attr-defined]: mypy thinks that phase is Phase type, while its actually PluginClass
```

File: tmt/steps/prepare/__init__.py [hunk 6]
```diff
@@ -223,7 +224,8 @@ def as_key(self) -> frozenset['tmt.base.DependencySimple']:
                     continue
 
                 collected_requires[guest].dependencies += tmt.base.assert_simple_dependencies(
-                    phase.essential_requires(),
+                    # mypy thinks that phase is Phase type, while its actually PluginClass
```
Comment (happz) [inline tmt/steps/prepare/__init__.py:227]:
```suggestion
                    # ignore[attr-defined]: mypy thinks that phase is Phase type, while its actually PluginClass
```

File: tmt/steps/prepare/__init__.py (hunk 4 above)
Comment (happz) [reply tmt/steps/prepare/__init__.py:190]:
```suggestion
                # ignore[attr-defined]: mypy doesn't seem to understand collection as `frozenset`
```

File: tmt/steps/prepare/__init__.py [hunk 7]
```diff
@@ -223,7 +224,8 @@ def as_key(self) -> frozenset['tmt.base.DependencySimple']:
                     continue
 
                 collected_requires[guest].dependencies += tmt.base.assert_simple_dependencies(
-                    phase.essential_requires(),
+                    # ignore[attr-defined]: mypy thinks that phase is Phase type, while its actually PluginClass
```
Comment (happz) [inline tmt/steps/prepare/__init__.py:227]:
```suggestion
                    # ignore[attr-defined]: mypy thinks that phase is Phase type, while its
                    # actually PluginClass
```

**Comment by happz** [comment]
Unrelated failure, merging.

**Pull request merged**